tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc

```diff
@@ -0,0 +1,447 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/compiler/xla/service/gpu/horizontal_fusion.h"
+
+#include <algorithm>
+
+#include "absl/container/flat_hash_set.h"
+#include "absl/types/span.h"
+#include "tensorflow/compiler/xla/layout_util.h"
+#include "tensorflow/compiler/xla/service/hlo_creation_utils.h"
+#include "tensorflow/core/util/env_var.h"
+
+namespace xla {
+namespace gpu {
+
+namespace {
+
+absl::InlinedVector<HloInstruction*, 2> GetOutputsOfFusion(
+    const HloInstruction& insn) {
+  CHECK(insn.opcode() == HloOpcode::kFusion);
+  auto root = insn.fused_expression_root();
+  if (root->opcode() != HloOpcode::kTuple) {
+    return {root};
+  } else {
+    return root->operands();
+  }
+}
+
+// Return the number of outputs of the fused computation.
+size_t GetOutputSizeOfFusion(const HloInstruction& insn) {
+  CHECK(insn.opcode() == HloOpcode::kFusion);
+  auto root = insn.fused_expression_root();
+  if (root->opcode() != HloOpcode::kTuple) {
+    return 1;
+  } else {
+    return ShapeUtil::TupleElementCount(root->shape());
+  }
+}
+
+class HorizontalFusionImpl {
+ public:
+  HorizontalFusionImpl(HloComputation* computation)
+      : computation_(computation) {}
+
+  ~HorizontalFusionImpl() {}
+
+  StatusOr<bool> Run();
+
+ private:
+  Status Fuse(absl::Span<HloInstruction*> fused_fusion_insns);
+
+  Status CreateFusedComputation(
+      absl::Span<HloInstruction*> fused_fusion_insns,
+      std::unique_ptr<HloComputation>* uniq_computation,
+      std::vector<HloInstruction*>* bound_operands);
+
+  // FusionCandidates collects profitable candidates for a given consumer
+  // instruction. GetNextSpanOfFusions() can then be iteratively invoked to
+  // acquire the next set of fusion candidates based on some heuristics.
+  class FusionCandidates {
+   public:
+    FusionCandidates(HloInstruction* consumer) : fusion_insns_(), pos_(0) {
+      Initialize(consumer);
+    }
+
+    // Get a span of fusions to be fused.
+    absl::Span<HloInstruction*> GetNextSpanOfFusions();
+
+   private:
+    void Initialize(HloInstruction*);
+
+    std::vector<HloInstruction*> fusion_insns_;
+    // pos_ points to the start position of next span.
+    size_t pos_;
+  };
+
+  HloComputation* computation_;
+};  // HorizontalFusionImpl
+
+void HorizontalFusionImpl::FusionCandidates::Initialize(
+    HloInstruction* consumer) {
+  // First, find out all fusion instructions. We will filter out
+  // unsupported/non-profitable cases below.
+  absl::flat_hash_set<HloInstruction*> fusion_insns;
+  for (auto opnd : consumer->operands()) {
+    auto predecessor = opnd->LatestNonGteAncestor();
+    if (predecessor->opcode() == HloOpcode::kFusion) {
```

Comment:
The primary reason is to address the main use case in my mind, i.e., training optimizers. If we see cases where fusing unfused HLO instructions are important, we can extend the capability.

Reply:
To be fair, this is inline with how fusion passes in XLA have been developed so far. Other passes treat fused and unfused ops differently too.
Sometimes I wish they wouldn't though, because there is no fundamental reason or benefit in doing so. The downside is that reasoning about optimized HLO ("Why do I see the fusions I see?") gets harder.

Reply:
Just to be clear, I agree with you. Extending this to unfused HLOs can be done later, if needed, and is worth a separate PR.

Reply:
See your point.